lib/netext/httpext/request.go

```diff
@@ -0,0 +1,372 @@
+		if preq.ResponseType == ResponseTypeNone {
+			_, err := io.Copy(ioutil.Discard, res.Body)
+			if err != nil && err != io.EOF {
+				resErr = err
+			}
+			resp.Body = nil
+		} else {
+			// Binary or string
+			buf := state.BPool.Get()
+			buf.Reset()
+			defer state.BPool.Put(buf)
+			_, err := io.Copy(buf, res.Body)
+			if err != nil && err != io.EOF {
+				resErr = err
+			}
+
+			switch preq.ResponseType {
+			case ResponseTypeText:
+				resp.Body = buf.String()
+			case ResponseTypeBinary:
+				resp.Body = buf.Bytes()
+			default:
+				resErr = fmt.Errorf("unknown responseType %s", preq.ResponseType)
+			}
+		}
+		_ = res.Body.Close()
+	}
+
+	trail := tracerTransport.GetTrail()
+
+	if trail.ConnRemoteAddr != nil {
+		remoteHost, remotePortStr, _ := net.SplitHostPort(trail.ConnRemoteAddr.String())
+		remotePort, _ := strconv.Atoi(remotePortStr)
+		resp.RemoteIP = remoteHost
+		resp.RemotePort = remotePort
+	}
+	resp.Timings = ResponseTimings{
+		Duration:       stats.D(trail.Duration),
+		Blocked:        stats.D(trail.Blocked),
+		Connecting:     stats.D(trail.Connecting),
+		TLSHandshaking: stats.D(trail.TLSHandshaking),
+		Sending:        stats.D(trail.Sending),
+		Waiting:        stats.D(trail.Waiting),
+		Receiving:      stats.D(trail.Receiving),
+	}
+
+	if resErr != nil {
+		resp.setError(resErr)
+	} else {
+		if preq.ActiveJar != nil {
+			if rc := res.Cookies(); len(rc) > 0 {
+				preq.ActiveJar.SetCookies(res.Request.URL, rc)
+			}
+		}
+
+		resp.URL = res.Request.URL.String()
+		resp.setStatusCode(res.StatusCode)
+		resp.Proto = res.Proto
+
+		if res.TLS != nil {
+			resp.setTLSInfo(res.TLS)
+		}
+
+		resp.Headers = make(map[string]string, len(res.Header))
+		for k, vs := range res.Header {
+			resp.Headers[k] = strings.Join(vs, ", ")
+		}
+
+		resCookies := res.Cookies()
+		resp.Cookies = make(map[string][]*HTTPCookie, len(resCookies))
+		for _, c := range resCookies {
+			resp.Cookies[c.Name] = append(resp.Cookies[c.Name], &HTTPCookie{
+				Name:     c.Name,
+				Value:    c.Value,
+				Domain:   c.Domain,
+				Path:     c.Path,
+				HTTPOnly: c.HttpOnly,
+				Secure:   c.Secure,
+				MaxAge:   c.MaxAge,
+				Expires:  c.Expires.UnixNano() / 1000000,
+			})
+		}
+	}
+
+	if resErr != nil {
+		// Do *not* log errors about the contex being cancelled.
+		select {
+		case <-ctx.Done():
+		default:
+			state.Logger.WithField("error", resErr).Warn("Request Failed")
+		}
+
+		if preq.Throw {
+			return nil, resErr
+		}
+	}
+
+	return resp, nil
+}
+
+// SetRequestCookies sets the cookies of the requests getting those cookies both from the jar and
+// from the reqCookies map. The Replace field of the HTTPRequestCookie will be taken into account
+func SetRequestCookies(req *http.Request, jar *cookiejar.Jar, reqCookies map[string]*HTTPRequestCookie) {
+	var replacedCookies = make(map[string]struct{})
+	for key, reqCookie := range reqCookies {
+		req.AddCookie(&http.Cookie{Name: key, Value: reqCookie.Value})
```

Comment:
I think this reverses the order of adding cookies from the previous `mergeCookies()` logic... Then cookiejar cookies were added first, before the cookies specified in the request params, while here the order is reversed. And it seems more logical to first emit the "global" (i.e. cookie jar) cookies, and then the request-specific ones. Though, considering that both the previous code and this one iterate over the cookie names in a semi-random order (due to the maps), I guess the point is kind of moot...

Reply:
as far as I know we are not really adhering to the [rfc6265 section 5.4 point 2](https://tools.ietf.org/html/rfc6265) although [cookiejar.Cookies](https://golang.org/src/net/http/cookiejar/jar.go?s=4642:4700#207) does. 
 Which probably ... makes this change to implement the rfc better as the cookies that will go first are the cookies for the specific URL that were provided. So I would argue that I (unintentionally) made us (more) compliant with the rfc ;)